torch/_inductor/ir.py

```diff
@@ -4542,6 +4544,52 @@ def __init__(
         self.kwargs = {} if kwargs is None else kwargs
         V.graph.warn_fallback(self.python_kernel_name)
 
+        # args that are aliased
+        self.alias_names = []
+        # args that are mutated AND returned from the op
+        self.mutation_names = []
+
+        if isinstance(self.op_overload, torch._ops.HigherOrderOperator):
+            # We assume here that HOPs with FallbackKernel are functional.
+            # This may not always be true! HOPs must individually opt-in to
+            # FallbackKernel, so please check this if you opt-in.
+            return
+
+        schema = self.op_overload._schema
+
+        # NOTE: [FallbackKernel supported operators]
+        # We only support three types of operators:
+        # - functional ops
+        # - view ops
+        # - inplace aten ops
+        # - mutating ops that are auto-functionalizable. That is,
+        # the operator may mutate any number of inputs, but its outputs
+        # may not alias any of the inputs.
+        #
+        # The unsupported cases usually do not show up here (because
+        # AOTAutograd functionalized them away); the only way for an in-place
+        # op to show up here is if a lowering or pass introduced it.
+        if torch._library.utils.is_inplace_aten_op(self.op_overload):
+            self.mutation_names.append(tensor_args[0].get_name())
+            return
+
+        if schema.is_mutable and not can_auto_functionalize(kernel):
+            raise NotImplementedError(
+                f"NYI: Can't generate FallbackKernel for {kernel}"
+            )
+
+        schema_args = schema.arguments
+        args, kwargs = self.unflatten_args(self.inputs, self.constant_args)
+        input_alias_sets = set()
+        for arg, info in zip(args, schema_args):
```

Comment:
I'm not sure that these two match up. schema_args will not unroll Tensor lists for example. could we add some assertions of some kind or another

Reply:
Yeah, I'll work through this more

Reply:
I think an assertion for isinstance(Tensor) is sufficient here, will put it in.

Reply:
What about `func(Tensor[] inp, tensor (a)!` ? In this case the schema and args will still be offset

Reply:
What about `func(Tensor[] inp, Tensor (a!) inp2)` ? Wont we still unflatten the inp list and have the mutation arg not be aligned ?

Reply:
@eellison I put in a bunch of safety checks, please let me know if that's what you were looking for.